astropy/visualization/lupton_rgb.py

```diff
@@ -53,6 +60,307 @@ def compute_intensity(image_r, image_g=None, image_b=None):
+
+        _Mapping.__init__(self, minimum=minimum, image=image)
+        self.maximum = maximum
+
+        if maximum is None:
+            self._range = None
+        else:
+            if maximum == minimum:
+                raise ValueError("minimum and maximum values must not be equal")
+            self._range = float(maximum - minimum)
+
+    def map_intensity_to_uint8(self, I):
+        # n.b. np.where can't and doesn't short-circuit
+        with np.errstate(invalid="ignore", divide="ignore"):
+            return np.where(
+                I <= 0,
+                0,
+                np.where(
+                    I >= self._range, self._uint8Max / I, self._uint8Max / self._range
+                ),
+            )
+
+
+class AsinhMapping(_Mapping):
+    """
+    A mapping for an asinh stretch (preserving colours independent of brightness).
+
+    x = asinh(Q (I - minimum)/stretch)/Q
+
+    This reduces to a linear stretch if Q == 0
+
+    See https://ui.adsabs.harvard.edu/abs/2004PASP..116..133L
+
+    Parameters
+    ----------
+    minimum : float
+        Intensity that should be mapped to black (a scalar or array for R, G, B).
+    stretch : float
+        The linear stretch of the image.
+    Q : float
+        The asinh softening parameter.
+    """
+
+    def __init__(self, minimum, stretch, Q=8):
+        _Mapping.__init__(self, minimum)
+
+        # 32bit floating point machine epsilon; sys.float_info.epsilon is 64bit
+        epsilon = 1.0 / 2**23
+        if abs(Q) < epsilon:
+            Q = 0.1
+        else:
+            Qmax = 1e10
+            if Q > Qmax:
+                Q = Qmax
+
+        frac = 0.1  # gradient estimated using frac*stretch is _slope
+        self._slope = frac * self._uint8Max / np.arcsinh(frac * Q)
+
+        self._soften = Q / float(stretch)
+
+    def map_intensity_to_uint8(self, I):
+        # n.b. np.where can't and doesn't short-circuit
+        with np.errstate(invalid="ignore", divide="ignore"):
+            return np.where(I <= 0, 0, np.arcsinh(I * self._soften) * self._slope / I)
+
+
+class AsinhZScaleMapping(AsinhMapping):
+    """
+    A mapping for an asinh stretch, estimating the linear stretch by zscale.
+
+    x = asinh(Q (I - z1)/(z2 - z1))/Q
+
+    Parameters
+    ----------
+    image1 : ndarray or a list of arrays
+        The image to analyse, or a list of 3 images to be converted to
+        an intensity image.
+    image2 : ndarray, optional
+        the second image to analyse (must be specified with image3).
+    image3 : ndarray, optional
+        the third image to analyse (must be specified with image2).
+    Q : float, optional
+        The asinh softening parameter. Default is 8.
+    pedestal : float or sequence(3), optional
+        The value, or array of 3 values, to subtract from the images; or None.
+
+    Notes
+    -----
+    pedestal, if not None, is removed from the images when calculating the
+    zscale stretch, and added back into Mapping.minimum[]
```

Comment:
I think things like `Mapping.minimum[]` in docstring might also be tripping the refs. There are auto ref magicks in several places. I will have to remember where.

Reply:
Hmm maybe not. I don't see it in

https://github.com/astropy/astropy/blob/77bb87d50a9944806312cefab2b96608482277a2/docs/conf.py#L511

Reply:
thanks, let's try this again !